lib/upload-manager.js

```diff
@@ -0,0 +1,287 @@
+/**
+ * @fileoverview Upload manager for large file uploads
+ */
+
+'use strict';
+
+// -----------------------------------------------------------------------------
+// Requirements
+// -----------------------------------------------------------------------------
+
+const EventEmitter = require('events').EventEmitter,
+	ReadableStream = require('stream').Readable,
+	crypto = require('crypto');
+
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+//@TODO: make into an option
+const RETRY_INTERVAL_MS = 1000;
+
+const DEFAULT_OPTIONS = {
+	parallelism: 4
+};
+
+class Chunk extends EventEmitter {
+
+	/**
+	 * Create a Chunk, representing a part of a file being uploaded
+	 * @param {BoxClient} client The Box SDK client
+	 * @param {string} sessionID The ID of the upload session the chunk belongs to
+	 * @param {?Buffer|string} chunk The chunk that was uploaded
+	 * @param {int} offset The bute offset within the file where this chunk begins
+	 * @param {int} totalSize The total size of the file this chunk belongs to
+	 */
+	constructor(client, sessionID, chunk, offset, totalSize) {
+
+		super();
+
+		this.client = client;
+		this.sessionID = sessionID;
+		this.chunk = chunk;
+		this.length = chunk.length;
+		this.offset = offset;
+		this.totalSize = totalSize;
+		this.data = null;
+		this.retry = null;
+		this.canceled = false;
+	}
+
+	/**
+	 * Get the final object representation of this chunk for the API
+	 * @returns {UploadPart} The chunk object
+	 */
+	getData() {
+
+		return this.data;
```

Comment:
No real reason, I guess.  The way I was storing the data changed a few times during development, so I used a method to maintain a decent interface.